### ISSUE TITLE
💬 RFC: Simplify and Standardize the approach for Pull Request and Merge Requests in Scaffolder Actions

### ISSUE THREAD

**Issue by benjdlambert** [backstage/backstage#22244]
### 🔖 Need

We have a large amount of actions that deal with upstream VCS providers like GitHub, Bitbucket, GitLab etc. There's the ability to create Pull Requests / Merge Requests from Branches, and these actions implement this in different ways. Some of the VCS providers implement standlone rest API's / GraphQL API's to be able to create Pull Requests or Merge Requests without checking out the code and operating with the Git Protocol, you can create commits with file changes through the API which in turn becomes a PR (looking at you GitHub).

This can lead to some issues like https://github.com/backstage/backstage/issues/17188, and sadly sometimes this approach is not available with the provider API's. (looking at you Bitbucket / GitLab).

Of course we want this functionality to be available for as many providers as possible, so thinking that theres chance to align the approach for all providers.

### 🎉 Proposal

We take advantage of the `isomorphic-git` library and prefer to use native git protocols instead of using the provider API's where this makes sense.

PR/MR actions will follow the following workflow:

- Clone the repository
- Create a Branch
- Make the required changes
  - Add/Remove/Edit files
- Create a commit
- Push the commit to the branch
- Use the provider specific API's to create a PR from the HEAD of the changed branch towards the base branch

We should be able to make this a non-breaking change, as we should have all the required permissions* and inputs required to make this a reality.

### 〽️ Alternatives

Open for other suggestions here, but we could just leave it as is today and accept the fact that bigger pull requests could fail because of rate limits, and different providers have different abilities.

### ❌ Risks

* This approach could be much slower on bigger repositories and for smaller changes. Thinking that we could lean towards falling back to this approach if there's a count of file changes above a certain threshold. Maybe we can just play this by ear though.
* Need to confirm that the user token permissions should be enough as is what happens today. Just want to make sure that we don't need to create forks of repositories, because that could be unfortunate.

### 👀 Have you spent some time to check if this RFC has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

**Comment by acierto** [comment]
@benjdlambert What do you think about including here also idempotency and aborting functionality?

**Comment by benjdlambert** [comment]
@acierto Probably a separate RFC to be honest - but for sure we should be thinking about this for all actions, although arguably these actions are the more important ones to make sure that they be idempotent.

**Comment by freben** [comment]
Does this have complexities in terms of auth? Ie passing down auth to isomorphic-git uniformly with what you pass to the native APIs

**Comment by benjdlambert** [comment]
I don't think that it should be an issue. We already use this approach for pushing new repositories with user `oauth` tokens. So I'm kind of feeling that it should be possible.

**Comment by benjdlambert** [comment]
Here's an example PR that implements basically something that I think we should do for all providers.

https://github.com/backstage/backstage/pull/21977

**Comment by afscrome** [comment]
For what it's worth, Azure devops plugins is using `isomorphic-git` for cloning & pushing.  (And then the ADO api for opening a pull request) - https://github.com/Parfuemerie-Douglas/scaffolder-backend-module-azure-repositories/blob/main/src/actions/run/cloneAzureRepo.ts .

**Comment by benjdlambert** [comment]
@afscrome this is exactly what I think we should be doing for all the providers and modules in `backstage/backstage` too :tada: Looks great!

**Comment by seanpoulter-foci** [comment]
Would Spotify be willing to help maintain `isomorphic-git` @benjdlambert? The tests need some TLC to get them stable.

**Comment by benjdlambert** [comment]
hey @seanpoulter-foci :wave:

that's an interesting question, of course I'd be willing to help out, however my schedule for the foreseeable future is pretty full and I don't have a lot of time outside of Backstage short term to help. 

Might have to get back to you later once things get a little more flexible :pray:

**Comment by Teo-ShaoWei** [comment]
When we say "Clone the repository", is it sufficient to clone just the latest main commit? E.g. with [isomorphhic-git's clone](https://isomorphic-git.org/docs/en/clone) setting `depth=1`, `singleBranch=true`, `ref=${configuredDefaultBranch}`.

**Comment by benjdlambert** [comment]
@Teo-ShaoWei I think that it could probably be parameter to change the behaviour, but the default should be a shallow clone for sure.

**Comment by Teo-ShaoWei** [comment]
Sorry, I meant I'll rather your implementation become the default, i.e. handle Git ops locally, over sending files through commit API to the VCS providers 👍

I'll need the GitLab ones to be fixed and standardized by the end of the year as we are going to use that, so I can help with that if this RFC gets accepted 💪

**Comment by seanpoulter-foci** [comment]
Please include removing file(s) in the RFC. There's a problems with the current scaffolder tracked in another issue. 

```diff
  PR/MR actions will follow the following workflow:

  - Clone the repository
  - Create a Branch
  - Make the required changes:
+    - Add files
+    - Modify files
+    - Remove files
  - Create a commit
```

**Comment by benjdlambert** [comment]
@seanpoulter-foci added 👍

**Comment by dotboris** [comment]
I wanted to chime in about commit signatures using GitHub apps. If we adopt native git protocols this presents a roadblock.

The only way for a GitHub app to be able to sign a commit is to use one of their APIs. You can't use GPG through git like you normally would. In fact, there's no where to upload a public GPG key for GitHub apps.

> Signature verification for bots will only work if the request is verified and authenticated as the GitHub App or bot and contains no custom author information, custom committer information, and no custom signature information, such as Commits API.
> Source: https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#signature-verification-for-bots

This paragraph seems to indicate that it might be possible to sign commits by clearing the author and committer information through git but it seems like that's not possible in practice: https://github.com/orgs/community/discussions/24664#discussioncomment-5084236

**Comment by benjdlambert** [comment]
@dotboris thanks for raising this - definitely a good callout.

I think what this most likely means is that we will probably have a separate action just for this special use case for GitHub and PGP Signing with an app instead, that creates the commits etc using the GitHub API. I'm not sure how this is is going to translate very well for Pull Requests, as it feels like we're just going to run into some of the issues mentioned already with that approach that we do today #17188.

Maybe we could turn back on throttling or something and accept that it's going to take a while to create them when using a GitHub App and for big pull requests. I'd rather have one approach that works for 99% of all use cases across all providers, rather than blocking it because of this one use case, and have another method.

**Comment by bbckr** [comment]
In case this is useful, I recently wrote a plugin that takes advantage of nodegit for all the git actions that you mentioned https://github.com/bbckr/backstage-plugins/tree/main/plugins/scaffolder-backend-module-git

It relies on the integration config for configuring oauth tokens so far for only gitlab and github, but can be used as a starting point. The usage of this is why I initially created issue #26845, since we have more control over how commits are pushed.

**Comment by mjramer** [comment]
Any update on this thread? Was planning to open a PR for an Azure DevOps PR scaffolder action with a near 1:1 match of the existing GitHub PR action
